elastic/beats
# [Autodiscover] Check if runner is already running before starting again

## What does this PR do?
This PR fixes runner reload so as not to start a new runner if a runner for the same configuration is already running. This can happen in Autodiscover if we have a container queued for termination and a new one with the very same configuration. This will lead into having 2 identical configurations in reload. The first one will be skipped but the second one will create new runner while the previous is still running. This is the tricky `if/else` block that cause this problem when we have 2 identical configurations: https://github.com/elastic/beats/blob/e99074029172a9c6d01f953005c3cdc2b58d6cb2/libbeat/cfgfile/list.go#L73

For more information check the related Discuss topic: https://discuss.elastic.co/t/multiple-monitoring-cycles-after-recreating-docker-image/231565/9

## Why is it important?
In case of autodiscovery catches a new start event will try to start a new runner without checking if a runner is already running. This will lead in overriding the in list of runner the old one with the new one without stoping the old one. The result will be to have 2 runners running (one will be orphan and untracked).

## Checklist

## How to test this PR locally
1. Enable autodiscover:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            contains:
              docker.container.image: prometheus
          config:
            - module: prometheus
              metricsets: ["collector"]
              hosts: "${data.host}:${data.port}"
```
2. Start Metricbeat: `./metricbeat -e -d "module,autodiscover"`
3. Start a container that matches the template using this docker-compose project: https://github.com/ChrsMark/docker-prometheus-playground
4. Edit the Prometheus service by adding a new label on it:
```
  prometheus:
    labels:
    - "some=Some"
```
5. Restart the service with `docker-compose up -d`
6. Verify that no new runner start:
There is no: `2020-05-15T08:25:01.563Z	DEBUG	[module]	module/wrapper.go:127	Starting Wrapper[name=prometheus, len(metricSetWrappers)=1]`
And there is: 
```
2020-05-21T07:47:11.487Z	DEBUG	[autodiscover]	cfgfile/list.go:62	Starting reload procedure, current runners: 1
2020-05-21T07:47:11.487Z	DEBUG	[autodiscover]	cfgfile/list.go:80	Start list: 0, Stop list: 0
```

## Related issues
Discuss: https://discuss.elastic.co/t/multiple-monitoring-cycles-after-recreating-docker-image/231565/9

This might solve https://github.com/elastic/beats/issues/12011 too.
